### ISSUE TITLE
npmjs.com/install.sh fails silently on clean Debian 9 install

### ISSUE THREAD

**Issue by strugee** [npm/npm#17875]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### How can the CLI team reproduce the problem?

1. Spin up a clean Debian 9 ("stretch") VM. I used a DigitalOcean droplet.
2. Log in via SSH
3. Run `apt update` and `apt upgrade`
4. `apt install nodejs nodejs-legacy curl`
5. `curl -L https://www.npmjs.com/install.sh > npm-install.sh`, because I don't do `curl | sh`
6. `bash npm-install.sh`

Notice how the install script claims that "it worked", but actually `/usr/lib/node_modules/npm` is a broken symlink to a deleted folder in `/tmp` (`/usr/bin/pump` points to where it should, and would definitely work if the symlink in `/usr/lib/node_modules` wasn't broken).

### supporting information:

 - `npm -v` prints: N/A
 - `node -v` prints: v4.8.2
 - `npm config get registry` prints: N/A
 - Windows, OS X/macOS, or Linux?: Linux
 - Network issues:
   - Geographic location where npm was run: reproduced worldwide (DigitalOcean San Francisco 1 datacenter, plus a private server Germany)
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by rDr4g0n** [comment]
I have the same issue. I am using a docker container from `debian:stable-slim`  

This command `curl -L https://npmjs.org/install.sh | sh` produces this output:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   185  100   185    0     0    419      0 --:--:-- --:--:-- --:--:--   418
100  6255  100  6255    0     0   6538      0 --:--:-- --:--:-- --:--:--  6538
tar=/usr/bin/tar
version:
tar (GNU tar) 1.29
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
install npm@latest
fetching: https://registry.npmjs.org/npm/-/npm-5.3.0.tgz
removed 411 packages in 6.859s
/usr/bin/npm -> /usr/lib/node_modules/npm/bin/npm-cli.js
/usr/bin/npx -> /usr/lib/node_modules/npm/bin/npx-cli.js
+ npm@5.3.0
added 1 package in 1.035s
It worked
```

However, `npm` is not found. following the symlink trail:
```
root@d60063b0f9bf:/usr/lib/node_modules# ls -ila /usr/bin/npm
65 lrwxrwxrwx 1 root root 38 Jul 25 02:27 /usr/bin/npm -> ../lib/node_modules/npm/bin/npm-cli.js
root@d60063b0f9bf:/usr/lib/node_modules# ls -ila /usr/lib/node_modules/npm    
67 lrwxrwxrwx 1 root root 27 Jul 25 02:27 /usr/lib/node_modules/npm -> ../../../tmp/npm.15/package
```
and the last path does not exist.

**Comment by mrjoops** [comment]
Same here.

As a workaround, line 263 of the `install.sh` script, I removed `  && rm -rf "$TMP" \` so the target of the symbolic link is not removed, and I could copy the content of the temp dir in `/usr/lib/node_modules/npm`. Not very clean by the way.

**Comment by cebe** [comment]
Similar issue, also on a fresh debian stretch:

```
# cat /etc/debian_version 
9.0
# nodejs --version
v4.8.2
# curl -L https://www.npmjs.com/install.sh | sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6255  100  6255    0     0  18837      0 --:--:-- --:--:-- --:--:-- 18897

npm cannot be installed without node.js.
Install node first, and then try again.

Maybe node is installed, but not in the PATH?
Note that running as sudo can change envs.
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Comment by cebe** [comment]
Seems the original issue is fixed by https://github.com/cebe/npm/commit/b58ec8eab3b4141e7f1b8b42d8cc24f716a804d8 but that commit is not included in the version of the file that is downloaded from https://www.npmjs.com/install.sh as of today.

**Comment by iarna** [comment]
@cebe Ah, that's a good insight! Yes, the install.sh is manually updated through dark rituals and we haven't managed to keep up a process for keeping it updated. (`install.sh` changes so rarely this doesn't come up often.)  I'll consult with the appropriate folks.

**Comment by robin-thoni** [comment]
When do you plan to deploy the new version? I also have some broken containers :)

**Comment by cebe** [comment]
@iarna thanks for the quick response and merging, however I'd have expected some discussion around the patch first :) It only solves part of the problem.

After some testing I figured out the npm installation still does not completely work as the `/usr/bin/npm` binary file is created to use the `node` binary instead of `nodejs`. This should also be adjusted by the install script.

**Comment by strugee** [comment]
@cebe this is a Debianism that no one really supports. Just install `nodejs-legacy` and note that the Technical Committee reversed their decision so in Debian Buster the binary will be called `node` again.

**Comment by cebe** [comment]
Yep, I learned that now, so I am not sure my PR should have been merged at all... @iarna please revert that commit, it is not a complete solution to the problem so it should not be released that way.

**Comment by strugee** [comment]
@iarna ping? This seems to still be broken :)

**Comment by gimite** [comment]
The issue still happens. I workarounded it by following https://github.com/npm/npm/pull/17626#issuecomment-316374409.

**Comment by strugee** [comment]
I just reproduced this on Ubuntu 18.04. Ping @iarna; this is still broken.
